jenkins/bootstrap.py

```diff
@@ -724,26 +728,52 @@ def gubernator_uri(paths):
     return job
 
 
-def setup_root(call, root, repo, branch, pull, ssh, git_cache, clean):
+def setup_root(call, root, repos, branches, pulls, ssh, git_cache, clean):
     """Create root dir, checkout repo and cd into resulting dir."""
-    logging.info(
-        'Check out %s at %s...',
-        os.path.join(root, repo or ''),
-        pull if pull else branch)
     if not os.path.exists(root):
         os.makedirs(root)
-    os.chdir(root)
-    if repo:
-        checkout(call, repo, branch, pull, ssh, git_cache, clean)
-    elif branch or pull:
-        raise ValueError('--branch and --pull require --repo', branch, pull)
-
-
-def bootstrap(
-    job, repo, branch, pull, root, upload, robot, timeout=0,
-    use_json=False, ssh=False, git_cache='', clean=False):
+    root_dir = os.path.realpath(root)
+    logging.info('Root: %s', root_dir)
+    if repos:
+        if branches:
+            if len(branches) != len(repos):
+                raise ValueError('--branch != --repo count', repos, branches)
+        if pulls:
+            if len(pulls) != len(repos):
+                raise ValueError('--pull != --repo count', repos, pulls)
+        for num, repo in enumerate(repos):
+            os.chdir(root_dir)
+            branch = branches and branches[num]
+            pull = pulls and pulls[num]
+            logging.info(
+                'Checkout: %s %s',
+                os.path.join(root_dir, repo or ''),
+                pull and pull or branch)
+            checkout(call, repo, branch, pull, ssh, git_cache, clean)
+        if len(repos) > 1:  # cd back into the primary repo
+            os.chdir(root_dir)
+            os.chdir(repos[0])
+    elif branches or pulls:
+        raise ValueError('--branch and --pull require --repo', branches, pulls)
+
+
+def bootstrap(args):
     """Clone repo at pull/branch into root and run job script."""
     # pylint: disable=too-many-locals,too-many-branches,too-many-statements
+    branches = args.branch
```

Comment:
It'd be clearer to keep these in args, and change the names to make sense as necessary. That also makes it easier to track where the values are coming from (command line vs derived).

For example:

    parser.add_argument('--pull', action='append', dest='pulls', help='PR, or list of PR:sha pairs like master:abcd,12:ef12,45:ff65')